File: appengine/cnpack-upgrade/cwstat.py
#!/usr/bin/env python
# coding=utf-8
#
# Copyright 2009 CnPack Team
#
# Author: Zhou Jingyu (zjy@cnpack.org)
#

import os
import cgi
import re
import datetime
import cndef
import wsgiref.handlers

from google.appengine.ext import db
from google.appengine.ext import webapp

class CnWizards(webapp.RequestHandler):

  def getReq(self, ident, default='none'):
    ret = self.request.get(ident)
    if ret == '':
      ret = default
    else:
      p = re.compile('[^a-zA-Z0-9_.\-]')
      if p.search(ret):
        ret = default
    return ret

  def incLog(self, ipaddr, now, ide, ver, code, lang):
    log = cndef.CWLogs(ipaddr=ipaddr, date=now, ide=ide, ver=ver, code=code, lang=lang)
    log.put()
    
  def incTemp(self, now, ide, ver, code, lang):
    rec = db.get(db.Key.from_path('CWConfig', 'D:Temp'))
    if (not rec) or (rec.value == '1'):
      Temp = cndef.CWTemp1(date=now, ide=ide, ver=ver, code=code, lang=lang)
    else:
      Temp = cndef.CWTemp2(date=now, ide=ide, ver=ver, code=code, lang=lang)
    Temp.put()

  def incCntHour(self, date, hour):
    rec = db.get(db.Key.from_path('CWCntHour', 'D:%s:%d' % (date, hour)))
    if not rec:
        rec = cndef.CWCntHour(date=date, hour=hour, count=1, key_name='D:%s:%d' % (date, hour))
    else:
        rec.count += 1
    rec.put()

  def incCntMonth(self, date_month):
    rec = db.get(db.Key.from_path('CWCntMonth', 'D:%s' % (date_month)))
    if not rec:
        rec = cndef.CWCntMonth(date=date_month, count=1, key_name='D:%s' % (date_month))
    else:
        rec.count += 1
    rec.put()

  def incCntUnion(self, date_month, ide, ver, code):
    rec = db.get(db.Key.from_path('CWCntUnion', 'D:%s:%s:%s:%s' % (date_month, ide, ver, code)))
    if not rec:
        rec = cndef.CWCntUnion(date=date_month, ide=ide, ver=ver, code=code, count=1, key_name='D:%s:%s:%s:%s' % (date_month, ide, ver, code))
    else:
        rec.count += 1
    rec.put()

  def outDatafile(self, fname):
    f = file(fname)
    while True:
        line = f.readline()
        if len(line) == 0: # Zero length indicates EOF
            break
        self.response.out.write(line)
    f.close() # close the file

  def get(self):
    # Use bin type to avoid utf-8 encoding error.
    self.response.headers["Content-Type"] = "application/octet-stream"
    self.outDatafile('cnwizards/update.ini')

    try:
      now = datetime.datetime.now() + datetime.timedelta(hours = 8)
      today = datetime.date(now.year, now.month, now.day)
      date_month = datetime.date(now.year, now.month, 1)
      alldate = datetime.date.min
  
      if self.request.get('month') == '1':
        db.run_in_transaction(self.incCntMonth, date_month)
  
      if self.request.get('manual') == '1':
        return
  
      db.run_in_transaction(self.incCntHour, today, now.hour)
      db.run_in_transaction(self.incCntHour, alldate, now.hour)

      ipaddr = os.environ['REMOTE_ADDR']
      ide = self.getReq('ide')
      ver = self.getReq('ver')
      code = cndef.country_code_by_addr(ipaddr)
      lang = self.getReq('langid')
  
      self.incLog(ipaddr, now, ide, ver, code, lang)
      self.incTemp(now, ide, ver, code, lang)
  
      db.run_in_transaction(self.incCntUnion, alldate, ide, ver, code)
      db.run_in_transaction(self.incCntUnion, date_month, ide, ver, code)
    except:
      return

def main():
  application = webapp.WSGIApplication([
    ('/cnwizards/', CnWizards)
], debug=True)

  wsgiref.handlers.CGIHandler().run(application)

if __name__ == '__main__':
  main()

File: appengine/cnpack-upgrade-php/include/cwstat.php
<?php

require_once ('common.php');
require_once ('db_mysql.inc');
require_once ('cndef.php');

function incLog($ipaddr, $now, $ide, $ver, $code, $lang)
{
  $db = new db("INSERT INTO cwlogs (ipaddr, date, ide, ver, code, lang) VALUES ('$ipaddr', '$now', '$ide', '$ver', '$code', '$lang')");
}

function incCntHour($date, $hour)
{
  $id = "D:$date:$hour";
  $db = new db("SELECT * FROM cwcnthour WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcnthour SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcnthour (id, date, hour, count) VALUES ('$id', '$date', '$hour', 1)");
}

function incCntMonth($date_month)
{
  $id = "D:$date_month";
  $db = new db("SELECT * FROM cwcntmonth WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntmonth SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntmonth (id, date, count) VALUES ('$id', '$date_month', 1)");
}

function incCntIde($date, $ide)
{
  $id = "D:$date:$ide";
  $db = new db("SELECT * FROM cwcntide WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntide SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntide (id, date, ide, count) VALUES ('$id', '$date', '$ide', 1)");
}

function incCntVer($date, $ver)
{
  $id = "D:$date:$ver";
  $db = new db("SELECT * FROM cwcntver WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntver SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntver (id, date, ver, count) VALUES ('$id', '$date', '$ver', 1)");
}

function incCntCountry($date, $code)
{
  $id = "D:$date:$code";
  $db = new db("SELECT * FROM cwcntcountry WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntcountry SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntcountry (id, date, code, count) VALUES ('$id', '$date', '$code', 1)");
}

function incCntLang($date, $lang)
{
  $id = "D:$date:$lang";
  $db = new db("SELECT * FROM cwcntlang WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntlang SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntlang (id, date, lang, count) VALUES ('$id', '$date', '$lang', 1)");
}

function incCntUnion($date_month, $ide, $ver, $code)
{
  $id = "D:$date_month:$ide:$ver:$code";
  $db = new db("SELECT * FROM cwcntunion WHERE id = '$id'");
  if ($db->next_record())
    $db->query("UPDATE cwcntunion SET count = count + 1 WHERE id = '$id'");
  else
    $db->query("INSERT INTO cwcntunion (id, date, ide, ver, code, count) VALUES ('$id', '$date_month', '$ide', '$ver', '$code', 1)");
}

function DoStat()
{
  date_default_timezone_set('PRC');
  $now = date("Y-m-d H:i:s");
  $today = date("Y-m-d");
  $date_month = date("Y-m-1");
  $alldate = date("2000-1-1");

  if (get_request('month') == '1')
    incCntMonth($date_month);

  if (get_request('manual') == '1')
    return;

  incCntHour($today, date("H"));
  incCntHour($alldate, date("H"));

  $ipaddr = get_request('ip');
  if (ip2long($ipaddr) == 0)
    $ipaddr = $_SERVER["REMOTE_ADDR"];
  
  $ide = get_request('ide', 'none');
  incCntIde($today, $ide);
  incCntIde($alldate, $ide);

  $ver = get_request('ver', 'none');
  incCntVer($today, $ver);
  incCntVer($alldate, $ver);

  $code = country_code_by_addr($ipaddr);
  incCntCountry($today, $code);
  incCntCountry($alldate, $code);

  $lang = get_request('langid', 'none');
  incCntLang($today, $lang);
  incCntLang($alldate, $lang);

  incLog($ipaddr, $now, $ide, $ver, $code, $lang);

  incCntUnion($alldate, $ide, $ver, $code);
  incCntUnion($date_month, $ide, $ver, $code);
}

DoStat();

?>